analyzer/java/com/engflow/bazel/invocation/analyzer/dataproviders/remoteexecution/CriticalPathQueuingDurationDataProvider.java

```diff
@@ -60,110 +62,174 @@ CriticalPathQueuingDuration getCriticalPathQueuingDuration()
     // the relevant name, and further filtering by time interval.
     // Given the matching event, find a queuing event with the same tid and pid that fits
     // within the time interval.
-    Set<CompleteEvent> criticalPathEventsInThreads = new HashSet<>();
+    HashMultimap<PidTidKey, CompleteEvent> criticalPathEventsInThreads = HashMultimap.create();
     if (bazelProfile.getCriticalPath().isEmpty()) {
       return new CriticalPathQueuingDuration(EMPTY_REASON);
     }
+    /*
+     * Key: critical action path event name.
+     * Value: the same critical action path, but in event form.
+     * Used to efficiently look up critical path events by their parsed name.
+     */
+    HashMap<String, CompleteEvent> cPathActionNameToSelfEvent = new HashMap<>();
+    /*
+     * Key: critical action path event name.
+     * Values: all profile 'action processing' events that matched by name with the critical path
+     *  counterpart.
+     * Used to keep track, for each critical path event, all the candidates that could be used to
+     * find its real process and thread IDs.
+     */
+    HashMultimap<String, CompleteEvent> cPathActionNameToEventCandidates = HashMultimap.create();
+    /*
+     * Key: PidTidKey (process id and thread id in record form) of the remote queuing event
+     * Value: The remote queuing event
+     * This is used to efficiently match between a critical path event with the remote queuing
+     * event, by the process and thread IDs.
+     */
+    HashMultimap<PidTidKey, CompleteEvent> remoteQueuingEvents = HashMultimap.create();
+
+    for (var cPathEvent : bazelProfile.getCriticalPath().get().getCompleteEvents()) {
```

Comment:
only do this when `bazelProfile.getCriticalPath().isPresent()`?


```suggestion
    bazelProfile.getCriticalPath().ifPresent(cp -> {
      for (var cPathEvent : cp.getCompleteEvents()) {
        ...
```

Reply:
This is actually already checked in https://github.com/EngFlow/bazel_invocation_analyzer/blob/cdb011deff8cd3bb72bf64d9b21890cce27db6c8/analyzer/java/com/engflow/bazel/invocation/analyzer/dataproviders/remoteexecution/CriticalPathQueuingDurationDataProvider.java#L66-L68

The `.isEmpty()` is a method from `Optional` that acts as the opposite of `.isPresent()`